### PR TITLE
Add VJP for concatenated.

### PR DESCRIPTION
**Pull request by pschuh**

### CODE REVIEW

File: stdlib/public/TensorFlow/Ops.swift [hunk 1]
```diff
@@ -663,24 +663,43 @@ public extension Tensor {
...
+}
+
+public extension Tensor {
   /// Concatenates tensors along the specified axis.
   /// - Precondition: The tensors must have the same dimensions, except for the
   ///   specified axis.
   /// - Precondition: The axis must be in the range `-rank..<rank`.
   @inlinable @inline(__always)
-  func concatenated(with other: Tensor, alongAxis axis: Int32) -> Tensor {
+  @differentiable(vjp:_concatenatedVJP)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: Are you sure this compiles?  The VJP is not defined in a generic context where `Scalar` conforms to `TensorFlowFloatingPoint`, and there's no `where` clause on the `@differentiable` attribute.

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/Ops.swift [hunk 2]
```diff
@@ -663,30 +663,50 @@ public extension Tensor {
...
-  /// Concatenates tensors along the first dimension.
-  /// - Precondition: The tensors must have the same shape, except for the
-  ///   leading dimension.
-  @inlinable @inline(__always)
-  func concatenated(with other: Tensor) -> Tensor {
-    return Raw.concatV2([self, other], axis: Tensor<Int32>(0))
-  }
 
+@inlinable @inline(__always)
+public func _splitV<T: TensorFlowScalar, Tidx: BinaryInteger & TensorFlowScalar>(
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: This helper function shouldn't be public. Is there no raw op for this? Use the `#tfop` directly in the VJP's body.

File: stdlib/public/TensorFlow/Ops.swift [hunk 3]
```diff
@@ -663,30 +663,50 @@ public extension Tensor {
...
+  @differentiable(vjp:_concatenatedVJP)
+  func concatenated(with other: Tensor, alongAxis axis: Int32 = 0) -> Tensor {
     return Raw.concatV2([self, other], axis: Tensor<Int32>(axis))
   }
 
+  @inlinable @inline(__always)
+  func _concatenatedVJP(with other: Tensor, alongAxis axis: Int32) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+    let idx = axis >= 0 ? axis : rank - axis
+    let splits = Tensor<Int32>([shape[idx], other.shape[idx]])
+    return (concatenated(with: other), {res in
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: * Define this in an `internal` extension where `Scalar : TensorFlowFloatingPoint`.
* Fit within 80 columns.

```suggestion
    return (concatenated(with: other), { res in
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 4]
```diff
@@ -663,30 +663,50 @@ public extension Tensor {
...
-  func concatenated(with other: Tensor, alongAxis axis: Int32) -> Tensor {
+  @differentiable(vjp:_concatenatedVJP)
+  func concatenated(with other: Tensor, alongAxis axis: Int32 = 0) -> Tensor {
     return Raw.concatV2([self, other], axis: Tensor<Int32>(axis))
   }
 
+  @inlinable @inline(__always)
+  func _concatenatedVJP(with other: Tensor, alongAxis axis: Int32) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+    let idx = axis >= 0 ? axis : rank - axis
+    let splits = Tensor<Int32>([shape[idx], other.shape[idx]])
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
    let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 5]
```diff
@@ -663,30 +663,50 @@ public extension Tensor {
...
+}
+
+public extension Tensor {
   /// Concatenates tensors along the specified axis.
   /// - Precondition: The tensors must have the same dimensions, except for the
   ///   specified axis.
   /// - Precondition: The axis must be in the range `-rank..<rank`.
   @inlinable @inline(__always)
-  func concatenated(with other: Tensor, alongAxis axis: Int32) -> Tensor {
+  @differentiable(vjp:_concatenatedVJP)
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
  @differentiable(vjp: _vjpConcatenated)
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 6]
```diff
@@ -663,30 +663,50 @@ public extension Tensor {
...
+  func concatenated(with other: Tensor, alongAxis axis: Int32 = 0) -> Tensor {
     return Raw.concatV2([self, other], axis: Tensor<Int32>(axis))
   }
 
+  @inlinable @inline(__always)
+  func _concatenatedVJP(with other: Tensor, alongAxis axis: Int32) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]: Please use `_vjpXXX` naming for consistency.
```suggestion
  func _vjpConcatenated(with other: Tensor, alongAxis axis: Int32) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
```

File: stdlib/public/TensorFlow/Ops.swift (hunk 2 above)
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
func _splitV<T: TensorFlowScalar, Tidx: BinaryInteger & TensorFlowScalar>(
```

**Review by dan-zheng — APPROVED** [review]
Could you please add a gradient test to `test/TensorFlowRuntime/tensor.swift`?

File: stdlib/public/TensorFlow/Ops.swift (hunk 2 above)
Comment (pschuh) [reply stdlib/public/TensorFlow/Ops.swift]: No raw op, returns a variable number of tensors...

File: stdlib/public/TensorFlow/Ops.swift (hunk 1 above)
Comment (pschuh) [reply stdlib/public/TensorFlow/Ops.swift]: Odd, this does compile though...

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/Ops.swift [hunk 7]
```diff
@@ -663,21 +663,15 @@ public extension Tensor {
...
-  }
 
+public extension Tensor {
   /// Concatenates tensors along the specified axis.
   /// - Precondition: The tensors must have the same dimensions, except for the
   ///   specified axis.
   /// - Precondition: The axis must be in the range `-rank..<rank`.
   @inlinable @inline(__always)
-  func concatenated(with other: Tensor, alongAxis axis: Int32) -> Tensor {
+  @differentiable(vjp:_vjpConcatenated where Scalar : TensorFlowFloatingPoint)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
  @differentiable(vjp: _vjpConcatenated where Scalar : TensorFlowFloatingPoint)
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 8]
```diff
@@ -687,11 +681,31 @@ public extension Tensor {
...
+  @differentiable(where Scalar : TensorFlowFloatingPoint)
   static func ++ (lhs: Tensor, rhs: Tensor) -> Tensor {
     return lhs.concatenated(with: rhs)
   }
 }
 
+internal extension Tensor where Scalar : TensorFlowFloatingPoint {
+  @inlinable @inline(__always)
+  func _vjpConcatenated(with other: Tensor, alongAxis axis: Int32)
+      -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
    -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 9]
```diff
@@ -687,11 +681,31 @@ public extension Tensor {
...
+    return (Raw.concatV2([self, other], axis: Tensor<Int32>(axis)), { result in
+        let ret: (TensorHandle<Scalar>, TensorHandle<Scalar>) = #tfop("SplitV",
+            result,
+            splits,
+            Tensor<Int32>(axis),
+            num_split: Int64(2),
+            T$dtype: Scalar.tensorFlowDataType,
+            Tlen$dtype: Int32.tensorFlowDataType)
+        return (Tensor(handle: ret.0), Tensor(handle: ret.1))
+      })
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
    })
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 10]
```diff
@@ -687,11 +681,31 @@ public extension Tensor {
...
 }
 
+internal extension Tensor where Scalar : TensorFlowFloatingPoint {
+  @inlinable @inline(__always)
+  func _vjpConcatenated(with other: Tensor, alongAxis axis: Int32)
+      -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+    let idx = axis >= 0 ? axis : rank - axis
+    let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
+    return (Raw.concatV2([self, other], axis: Tensor<Int32>(axis)), { result in
+        let ret: (TensorHandle<Scalar>, TensorHandle<Scalar>) = #tfop("SplitV",
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: Indent by 2

File: stdlib/public/TensorFlow/Ops.swift [hunk 11]
```diff
@@ -687,11 +681,31 @@ public extension Tensor {
...
   }
 }
 
+internal extension Tensor where Scalar : TensorFlowFloatingPoint {
+  @inlinable @inline(__always)
+  func _vjpConcatenated(with other: Tensor, alongAxis axis: Int32)
+      -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+    let idx = axis >= 0 ? axis : rank - axis
+    let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
+    return (Raw.concatV2([self, other], axis: Tensor<Int32>(axis)), { result in
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: Indent by 2

**Review by rxwei — APPROVED** [review]

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Pull request merged**